libs/execution_context/ebpf_maps.c

```diff
@@ -439,6 +445,311 @@ _next_hash_map_key(_In_ ebpf_core_map_t* map, _In_ const uint8_t* previous_key,
     return result;
 }
 
+static ebpf_core_map_t*
+_create_hash_map_per_cpu(_In_ const ebpf_map_definition_t* map_definition)
+{
+    ebpf_result_t retval;
+    size_t map_size = sizeof(ebpf_core_map_t);
+    ebpf_core_map_t* map = NULL;
+    uint32_t cpu_count;
+    uint32_t index;
+    size_t per_cpu_size = 0;
+    ebpf_core_map_per_cpu_t* per_cpu = NULL;
+
+    ebpf_get_cpu_count(&cpu_count);
+
+    retval = ebpf_safe_size_t_multiply(sizeof(ebpf_hash_table_t*), cpu_count, &per_cpu_size);
+    if (retval != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    retval = ebpf_safe_size_t_add(EBPF_OFFSET_OF(ebpf_core_map_per_cpu_t, data), per_cpu_size, &per_cpu_size);
+    if (retval != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    retval = ebpf_safe_size_t_add(per_cpu_size, map_size, &map_size);
+    if (retval != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    map = ebpf_allocate(map_size);
+    if (map == NULL) {
+        retval = EBPF_NO_MEMORY;
+        goto Done;
+    }
+
+    map->ebpf_map_definition = *map_definition;
+    map->data = (uint8_t*)(map + 1);
+    per_cpu = (ebpf_core_map_per_cpu_t*)map->data;
+
+    for (index = 0; index < cpu_count; index++) {
+        ebpf_hash_table_t** tables = (ebpf_hash_table_t**)&per_cpu->data;
+        retval = ebpf_hash_table_create(
+            tables + index,
+            ebpf_epoch_allocate,
+            ebpf_epoch_free,
+            map->ebpf_map_definition.key_size,
+            map->ebpf_map_definition.value_size,
+            NULL);
+        if (retval != EBPF_SUCCESS) {
+            goto Done;
+        }
+    }
+
+    map->data = (uint8_t*)per_cpu;
+    per_cpu->count = cpu_count;
+
+    ebpf_lock_create(&map->lock);
+    retval = EBPF_SUCCESS;
+
+Done:
+    if (retval != EBPF_SUCCESS) {
+        if (per_cpu) {
+            for (index = 0; index < per_cpu->count; index++) {
+                ebpf_hash_table_t** tables = (ebpf_hash_table_t**)&per_cpu->data;
+                ebpf_hash_table_destroy(tables[index]);
+            }
+        }
+        ebpf_free(map);
+        map = NULL;
+    }
+    return map;
+}
+
+static void
+_delete_hash_map_per_cpu(_In_ ebpf_core_map_t* map)
```

Comment:
this name looks like a misnomer.  If I understand right, there is one "map" with per-cpu data, not a separate map per fd
```suggestion
_delete_per_cpu_hash_map(_In_ ebpf_core_map_t* map)
```